File: HW/HW8/Main.py
class Solution:
    def longestPalindrome(self, s: str) -> int:
        from collections import Counter 
        count = Counter(s)
        palindrome_len = 0
        odd_exists = False
        
        for value in count.values():
            if value % 2 == 0:
                palindrome_len += value 
            else: 
                palindrome_len += value - 1 
                odd_exists = True
        if odd_exists:
            palindrome_len += 1
        return palindrome_len
        
scores = Solution()
print(scores.longestPalindrome("abccccdd"))
print(scores.longestPalindrome("speediskey"))

File: HW/HW8/Main.java
import java.util.HashMap;
public class Main
{
    public static int longestPalindrome(String s) {
        HashMap<Character, Integer> counters = new HashMap<>();
        for(char c : s.toCharArray())
            counters.put(c, counters.getOrDefault(c, 0)+1);
        int palindromeLength = 0;
        boolean couldHaveMidChar = false;
        for(Integer letterCount: counters.values()){
            if(letterCount%2 == 0)
                palindromeLength += letterCount;
            else {
                palindromeLength += letterCount/2*2;
                couldHaveMidChar = true;
            }
        }
        return couldHaveMidChar ? palindromeLength+1 : palindromeLength;
    }
    
    public static void main(String [] args) {
        System.out.println(longestPalindrome("abccccdd"));
        System.out.println(longestPalindrome("speediskey"));
    }
}